File: apps/web/src/api/effectRecovery.ts
import type { ApiClient, ApiOutcome } from './client'

export type DeliveryPhase = 'OPEN' | 'CLOSED_UNUSED' | 'UNCONFIRMED' | 'DELIVERY_RECORD_MISSING' | 'RESPONSE_RETAINED'
export interface EffectDelivery {
  readonly permitId: string
  readonly actionId: string
  readonly attemptId: string
  readonly actionSequence: number
  readonly action: string
  readonly actionResult: 'SUCCEEDED' | 'FAILED' | 'AMBIGUOUS' | null
  readonly phase: DeliveryPhase
  readonly grantedAt: string
  readonly expiresAt: string
  readonly consumedAt: string | null
  readonly responseRecordedAt: string | null
  readonly requestDigest: string | null
  readonly responseDigest: string | null
  readonly responseStatus: number | null
  readonly actionResultAt: string | null
  readonly leaseId: string
  readonly leaseReleasedAt: string | null
  readonly stopAcknowledgedAt: string | null
  readonly runnerQuarantined: boolean
  readonly endpointState: string | null
  readonly endpointCleanupConfirmed: boolean | null
  readonly requiresInvestigation: boolean
}
export interface EffectRecovery {
  readonly runId: string
  readonly observedAt: string
  readonly runStatus: string
  readonly runOutcome: string
  readonly runQuarantined: boolean
  readonly items: readonly EffectDelivery[]
  readonly nextCursor: string | null
  readonly providesRetryAuthority: false
  readonly providesResetAuthority: false
  readonly meaning: 'FORM_TRANSPORT_HISTORY_NOT_EFFECT_PROOF'
}

export const RECOVERY_PAGE_SIZE = 20
const object = (v: unknown): v is Record<string, unknown> => typeof v === 'object' && v !== null && !Array.isArray(v)
const text = (v: unknown): v is string => typeof v === 'string' && v.length > 0
const uuid = (v: unknown): v is string => typeof v === 'string' && /^[0-9a-f]{8}(-[0-9a-f]{4}){3}-[0-9a-f]{12}$/.test(v)
const hash = (v: unknown): boolean => typeof v === 'string' && /^[0-9a-f]{64}$/.test(v)
const time = (v: unknown): boolean => text(v) && /T.*(?:Z|[+-]\d{2}:\d{2})$/.test(v) && Number.isFinite(Date.parse(v))
const nullable = (v: unknown, check: (v: unknown) => boolean): boolean => v === null || check(v)

/** Display validation, not cryptographic verification or authority to repeat an effect. */
export function parseEffectRecovery(value: unknown, runId: string, after: string | null): EffectRecovery | null {
  if (!object(value) || value.runId !== runId || !uuid(value.runId) || !time(value.observedAt) ||
      !text(value.runStatus) || !text(value.runOutcome) || typeof value.runQuarantined !== 'boolean' ||
      value.meaning !== 'FORM_TRANSPORT_HISTORY_NOT_EFFECT_PROOF' || value.providesRetryAuthority !== false ||
      value.providesResetAuthority !== false || !nullable(value.nextCursor, uuid) ||
      !Array.isArray(value.items) || value.items.length > RECOVERY_PAGE_SIZE) return null
  let previous = after
  for (const item of value.items) {
    if (!object(item) || !uuid(item.permitId) || (previous !== null && item.permitId <= previous) ||
        !uuid(item.actionId) || !uuid(item.attemptId) || !uuid(item.leaseId) ||
        !Number.isSafeInteger(item.actionSequence) || (item.actionSequence as number) < 1 || !text(item.action) ||
        !nullable(item.actionResult, (v) => text(v) && ['SUCCEEDED', 'FAILED', 'AMBIGUOUS'].includes(v)) ||
        !text(item.phase) || !['OPEN', 'CLOSED_UNUSED', 'UNCONFIRMED', 'DELIVERY_RECORD_MISSING', 'RESPONSE_RETAINED'].includes(item.phase) ||
        !time(item.grantedAt) || !time(item.expiresAt) ||
        !['consumedAt', 'responseRecordedAt', 'actionResultAt', 'leaseReleasedAt', 'stopAcknowledgedAt'].every((key) => nullable(item[key], time)) ||
        !nullable(item.requestDigest, hash) || !nullable(item.responseDigest, hash) ||
        !nullable(item.responseStatus, (v) => typeof v === 'number' && [200, 201, 404, 409, 422].includes(v)) ||
        typeof item.runnerQuarantined !== 'boolean' || typeof item.requiresInvestigation !== 'boolean' ||
        !nullable(item.endpointState, text) || !nullable(item.endpointCleanupConfirmed, (v) => typeof v === 'boolean')) return null
    const consumed = item.phase !== 'OPEN' && item.phase !== 'CLOSED_UNUSED'
    const response = item.phase === 'RESPONSE_RETAINED'
    const request = response || item.phase === 'UNCONFIRMED'
    if (consumed !== (item.consumedAt !== null) || request !== (item.requestDigest !== null) ||
        response !== (item.responseDigest !== null) || response !== (item.responseStatus !== null) ||
        response !== (item.responseRecordedAt !== null) ||
        (item.actionResult !== null) !== (item.actionResultAt !== null) ||
        item.requiresInvestigation !== (['UNCONFIRMED', 'DELIVERY_RECORD_MISSING'].includes(item.phase) || item.actionResult === 'AMBIGUOUS')) return null
    previous = item.permitId
  }
  if (value.nextCursor !== null && (value.items.length !== RECOVERY_PAGE_SIZE || value.nextCursor !== previous)) return null
  return value as unknown as EffectRecovery
}

export async function readEffectRecovery(
  client: ApiClient, workspaceId: string, runId: string, after: string | null, signal: AbortSignal,
): Promise<ApiOutcome<EffectRecovery>> {
  const query = new URLSearchParams({ limit: String(RECOVERY_PAGE_SIZE) })
  if (after !== null) query.set('after', after)
  const response = await client.request<unknown>(
    `/v1/workspaces/${encodeURIComponent(workspaceId)}/runs/${encodeURIComponent(runId)}/effect-deliveries?${query}`, { signal },
  )
  if (response.kind !== 'ok' && response.kind !== 'accepted') return response
  const value = parseEffectRecovery(response.value, runId, after)
  if (value === null) return { kind: 'problem', problem: {
    code: 'UNRECOGNISED', title: 'Transport history cannot be displayed', status: 502, requestId: null,
    detail: 'The response is malformed, unsupported or does not match this run and page. No effect or recovery permission has been inferred.',
  } }
  return response.kind === 'accepted' ? { kind: 'accepted', value } : { ...response, value }
}

File: packages/persistence/src/accessforge_persistence/effect_recovery.py
"""Bounded historical form-transport inspection; no dispatch, reset or outcome inference."""

from __future__ import annotations

from typing import Any
from uuid import UUID

import psycopg

from accessforge_domain.timestamps import to_rfc3339_utc

from . import candidate_effects


class Refused(Exception):
    pass


def _time(value: Any) -> str | None:
    return None if value is None else to_rfc3339_utc(value)


def _item(row: dict[str, Any]) -> dict[str, Any]:
    try:
        permit = candidate_effects._view(row)
        grant = row["grant_payload"]
        if (
            grant["runId"] != str(row["requested_run_id"])
            or grant["attemptId"] != str(row["action_attempt_id"])
            or grant["leaseId"] != str(row["action_lease_id"])
            or grant["leaseEpoch"] != row["action_epoch"]
        ):
            raise Refused("original permission/action binding differs")
        if row["consumed_at"] is None:
            if row["delivery_id"] is not None:
                raise Refused("unconsumed permission has an inconsistent delivery record")
            phase = "OPEN" if row["expires_at"] > row["snapshot_at"] else "CLOSED_UNUSED"
        elif row["delivery_id"] is None:
            phase = "DELIVERY_RECORD_MISSING"
        elif row["response_digest"] is None:
            phase = "UNCONFIRMED"
        else:
            if (
                row["response_status"] not in (200, 201, 404, 409, 422)
                or row["responded_at"] is None
            ):
                raise Refused("retained transport response is inconsistent")
            phase = "RESPONSE_RETAINED"
        return {
            "permitId": permit["permitId"],
            "actionId": permit["actionId"],
            "attemptId": str(row["action_attempt_id"]),
            "actionSequence": row["action_sequence"],
            "action": row["action"],
            "actionResult": row["result_status"],
            "phase": phase,
            "grantedAt": _time(row["granted_at"]),
            "expiresAt": permit["expiresAt"],
            "consumedAt": _time(row["consumed_at"]),
            "responseRecordedAt": _time(row["responded_at"]),
            "requestDigest": row["request_digest"],
            "responseDigest": row["response_digest"],
            "responseStatus": row["response_status"],
            "actionResultAt": _time(row["result_at"]),
            "leaseId": str(row["action_lease_id"]),
            "leaseReleasedAt": _time(row["released_at"]),
            "stopAcknowledgedAt": _time(row["stop_acknowledged_at"]),
            "runnerQuarantined": row["runner_quarantined_at"] is not None,
            "endpointState": row["endpoint_state"],
            "endpointCleanupConfirmed": row["cleanup_confirmed"],
            "requiresInvestigation": phase in {"UNCONFIRMED", "DELIVERY_RECORD_MISSING"}
            or row["result_status"] == "AMBIGUOUS",
        }
    except (KeyError, TypeError, ValueError, candidate_effects.Refused) as exc:
        raise Refused("original form-transport history integrity unavailable") from exc


def read(
    conn: psycopg.Connection[Any],
    *,
    run_id: str,
    limit: int = 50,
    after: str | None = None,
) -> dict[str, Any] | None:
    """One statement snapshot. Cursor orders original permit UUIDs, not event chronology.

    Workspace RLS and the caller's evidence-read permission are required. Empty history does not
    prove zero application effects. This works after session expiry without renewing authority.
    """
    if type(limit) is not int or not 1 <= limit <= 100:
        raise Refused("recovery report limit must be between 1 and 100")
    try:
        run = str(UUID(run_id))
        cursor = None if after is None else str(UUID(after))
    except (ValueError, TypeError, AttributeError) as exc:
        raise Refused("recovery report identity unavailable") from exc
    rows = conn.execute(
        "SELECT r.id AS requested_run_id,r.status AS run_status,r.outcome AS run_outcome,"
        "r.quarantined AS run_quarantined,statement_timestamp() AS snapshot_at,page.* "
        "FROM run r LEFT JOIN LATERAL ("
        " SELECT p.*,a.attempt_id AS action_attempt_id,a.lease_id AS action_lease_id,"
        " a.epoch AS action_epoch,a.action_sequence,a.action,a.result_status,a.result_at,"
        " l.released_at,l.stop_acknowledged_at,n.quarantined_at AS runner_quarantined_at,"
        " e.state AS endpoint_state,e.cleanup_confirmed,d.permit_id AS delivery_id,"
        " d.request_digest,d.response_digest,d.response_status,d.responded_at "
        " FROM candidate_action_effect_permit p JOIN runner_action a ON a.id=p.action_id "
        " JOIN desktop_lease l ON l.id=a.lease_id JOIN runner n ON n.id=l.runner_id "
        " LEFT JOIN candidate_endpoint e ON e.attempt_id=p.regression_attempt_id "
        " LEFT JOIN candidate_effect_delivery d ON d.permit_id=p.id "
        " WHERE a.run_id=r.id AND (%s::uuid IS NULL OR p.id>%s::uuid) "
        " ORDER BY p.id LIMIT %s"
        ") page ON true WHERE r.id=%s AND r.workspace_id=current_workspace_id() ORDER BY page.id",
        (cursor, cursor, limit + 1, run),
    ).fetchall()
    if not rows:
        return None
    root = rows[0]
    retained = [row for row in rows if row["id"] is not None]
    page = retained[:limit]
    return {
        "runId": run,
        "observedAt": _time(root["snapshot_at"]),
        "runStatus": root["run_status"],
        "runOutcome": root["run_outcome"],
        "runQuarantined": root["run_quarantined"],
        "items": [_item(row) for row in page],
        "nextCursor": str(page[-1]["id"]) if len(retained) > limit else None,
        "providesRetryAuthority": False,
        "providesResetAuthority": False,
        "meaning": "FORM_TRANSPORT_HISTORY_NOT_EFFECT_PROOF",
    }
